src/pykeen/nn/emb.py

```diff
@@ -1068,3 +1070,133 @@ def forward(
+
+        :param k:
+            the number of tokens for each entity.
+        """
+        if triples_factory.create_inverse_triples:
+            raise NotImplementedError("Triples factories with inverse triples are currently not supported.")
+
+        # create token representations
+        num_tokens = 2 * triples_factory.num_relations + 1
+        if isinstance(token_representation, EmbeddingSpecification):
+            token_representation = token_representation.make(
+                num_embeddings=num_tokens,  # normal relations + inverse relations + padding
+            )
+        else:
+            if token_representation.max_id != num_tokens:
+                raise ValueError(
+                    f"If a pre-instantiated representation is provided, it has to have 2 * num_relations + 1 = "
+                    f"{num_tokens} representations, but has {token_representation.max_id}",
+                )
+
+        # super init; has to happen *before* any parameter or buffer is assigned
+        super().__init__(max_id=triples_factory.num_entities, shape=shape or token_representation.shape)
+
+        # normalize aggregation
+        if aggregation is None:
+            aggregation = torch.mean
+        self.aggregation = aggregation
+
+        # assign module
+        self.tokens = token_representation
+
+        # tokenize: represent entities by bag of relations
+        h, r, t = triples_factory.mapped_triples.t()
+
+        # collect candidates
+        e2r = defaultdict(set)
+        for e, r in (
+            torch.cat(
+                [
+                    torch.stack([h, r], dim=1),
+                    torch.stack([t, r + triples_factory.num_relations], dim=1),
+                ],
+                dim=0,
+            )
+            .unique(dim=0)
+            .tolist()
+        ):
+            e2r[e].add(r)
+
+        # randomly sample with replacement k relations for each entity
+        assignment = torch.full(
+            size=(triples_factory.num_entities, k),
+            dtype=torch.long,
+            fill_value=2 * triples_factory.num_relations,  # padding ID
+        )
+        for e, rs in e2r.items():
+            assignment[e] = self._sample(torch.as_tensor(data=list(rs), dtype=torch.long), k)
+        self.register_buffer(name="assignment", tensor=assignment)
+
+    @staticmethod
+    def _sample(rs: torch.LongTensor, k: int) -> torch.LongTensor:
+        return rs[torch.randint(high=rs.numel(), size=(k,))]
+
+    def forward(
+        self,
+        indices: Optional[int] = None,
+    ) -> torch.FloatTensor:  # noqa: D102
+        # get token IDs, shape: (*, k)
+        token_ids = self.assignment
+        if indices is not None:
+            token_ids = token_ids[indices]
+
+        # lookup token representations, shape: (*, k, d)
+        x = self.tokens(token_ids)
+
+        # aggregate
+        x = self.aggregation(x, -2)
```

Comment:
Add docs on what the `-2` means (typically, the dimension)

Reply:
I think we should just get rid of it (by applying functools.partial to torch.mean)

Reply:
but how should someone bring their own function then?